### ISSUE TITLE
[BUG] stochastic parameter shift rule does not work with hardware Hamiltonians (transmon / rydberg)

### ISSUE THREAD

**Issue by Qottmann** [PennyLaneAI/pennylane#4177]
### Expected behavior

[qml.gradients.stoch_pulse_grad](https://docs.pennylane.ai/en/latest/code/api/pennylane.gradients.stoch_pulse_grad.html) can handle pulse Hamiltonians

### Actual behavior

It currently cant.

### Additional information

The `HardwareHamiltonian` has drive terms $\Omega(p, t) (\cos(\phi) \sigma_x - \sin(\phi) \sigma_y)$, in particular two separate terms $\Omega(p, t) \cos(\phi) \sigma_x$ and $-\Omega(p, t) \sin(\phi) \sigma_y$ with shared parameters $p$. In order for the user not to have to pass the same argument twice, we take care of this with a reordering function https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/hardware_hamiltonian.py#L237

This however is currently not being taken into account in the stochastic pulse gradient transform. I think the problematic line is https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/gradients/pulse_gradient.py#L564

### Source code

```shell
# values from https://arxiv.org/pdf/2203.06818.pdf
qubit_freq = [4.8080, 4.8333]
connections = [(0, 1)]
coupling = 0.01831
wires = [0, 1]
n_wires = len(wires)
timespan = 20.

def amp(p, t):
    return jnp.polyval(p, t)

H = qml.pulse.transmon_interaction(qubit_freq, connections, coupling, wires)
H += qml.pulse.transmon_drive(amp, 0., qubit_freq[0], wires=[0])
H += qml.pulse.transmon_drive(amp, 0., qubit_freq[0], wires=[1])

atol=1e-5
num_split_times = 20

dev = qml.device("default.qubit.jax", wires=n_wires)

@qml.qnode(dev, interface="jax", diff_method=qml.gradients.stoch_pulse_grad, num_split_times=num_split_times)#, use_broadcasting=True)
def cost(params):
    qml.evolve(H, atol=atol)(params, t=timespan)
    return qml.expval(qml.PauliZ(0))

params = (jnp.ones(3), jnp.ones(3))
jax.grad(cost)(params)
```

### Tracebacks

```shell
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[21], line 1
----> 1 jax.grad(cost)(params)

    [... skipping hidden 10 frame]

File ~/Qottmann/Xanadu/pennylane/pennylane/qnode.py:883, in QNode.__call__(self, *args, **kwargs)
    881     self.execute_kwargs.pop("mode")
    882 # pylint: disable=unexpected-keyword-arg
--> 883 res = qml.execute(
    884     [self.tape],
    885     device=self.device,
    886     gradient_fn=self.gradient_fn,
    887     interface=self.interface,
    888     gradient_kwargs=self.gradient_kwargs,
    889     override_shots=override_shots,
    890     **self.execute_kwargs,
    891 )
    893 res = res[0]
    895 if old_interface == "auto":

File ~/Qottmann/Xanadu/pennylane/pennylane/interfaces/execution.py:573, in execute(tapes, device, gradient_fn, interface, grad_on_execution, gradient_kwargs, cache, cachesize, max_diff, override_shots, expand_fn, max_expansion, device_batch_transform)
    570     elif mapped_interface == "jax":
    571         _execute = _get_jax_execute_fn(interface, tapes)
--> 573     res = _execute(
    574         tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n=1, max_diff=max_diff
    575     )
    577 except ImportError as e:
    578     raise qml.QuantumFunctionError(
    579         f"{mapped_interface} not found. Please install the latest "
    580         f"version of {mapped_interface} to enable the '{mapped_interface}' interface."
    581     ) from e

File ~/Qottmann/Xanadu/pennylane/pennylane/interfaces/jax.py:416, in execute(tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n, max_diff)
    407     return _execute_fwd(
    408         parameters,
    409         tapes,
   (...)
    412         _n=_n,
    413     )
    415 # PennyLane backward execution
--> 416 return _execute_bwd(
    417     parameters,
    418     tapes,
    419     device,
    420     execute_fn,
    421     gradient_fn,
    422     gradient_kwargs,
    423     _n=_n,
    424     max_diff=max_diff,
    425 )

File ~/Qottmann/Xanadu/pennylane/pennylane/interfaces/jax.py:504, in _execute_bwd(params, tapes, device, execute_fn, gradient_fn, gradient_kwargs, _n, max_diff)
    500         jvps = _compute_jvps(jacs, tangents[0], multi_measurements)
    502     return res, jvps
--> 504 return execute_wrapper(params)

    [... skipping hidden 5 frame]

File ~/Qottmann/Xanadu/pennylane/pennylane/interfaces/jax.py:475, in _execute_bwd..execute_wrapper_jvp(primals, tangents)
    470 _kwargs = {
    471     "reduction": "append",
    472     "gradient_kwargs": gradient_kwargs,
    473 }
    474 if at_max_diff:
--> 475     jvp_tapes, processing_fn = qml.gradients.batch_jvp(*_args, **_kwargs)
    476     jvps = processing_fn(execute_fn(jvp_tapes)[0])
    477 else:

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/jvp.py:426, in batch_jvp(tapes, tangents, gradient_fn, shots, reduction, gradient_kwargs)
    424 # Loop through the tapes and dys vector
    425 for tape, tangent in zip(tapes, tangents):
--> 426     g_tapes, fn = jvp(tape, tangent, gradient_fn, shots, gradient_kwargs)
    428     reshape_info.append(len(g_tapes))
    429     processing_fns.append(fn)

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/jvp.py:326, in jvp(tape, tangent, gradient_fn, shots, gradient_kwargs)
    324 if shots is None:
    325     shots = tape.shots
--> 326 gradient_tapes, fn = gradient_fn(tape, shots=shots, **gradient_kwargs)
    328 def processing_fn(results):
    329     # postprocess results to compute the Jacobian
    330     jac = fn(results)

File ~/Qottmann/Xanadu/pennylane/pennylane/transforms/batch_transform.py:362, in batch_transform.__call__(self, *targs, **tkwargs)
    357     return self._device_wrapper(*targs, **tkwargs)(qnode)
    359 if isinstance(qnode, qml.tape.QuantumScript):
    360     # Input is a quantum tape.
    361     # tapes, fn = some_transform(tape, *transform_args)
--> 362     return self._tape_wrapper(*targs, **tkwargs)(qnode)
    364 if isinstance(qnode, (qml.QNode, qml.ExpvalCost)):
    365     # Input is a QNode:
    366     # result = some_transform(qnode, *transform_args)(*qnode_args)
    367     wrapper = self.qnode_wrapper(qnode, targs, tkwargs)

File ~/Qottmann/Xanadu/pennylane/pennylane/transforms/batch_transform.py:455, in batch_transform._tape_wrapper..(tape)
    454 def _tape_wrapper(self, *targs, **tkwargs):
--> 455     return lambda tape: self.construct(tape, *targs, **tkwargs)

File ~/Qottmann/Xanadu/pennylane/pennylane/transforms/batch_transform.py:437, in batch_transform.construct(self, tape, *targs, **tkwargs)
    435 if argnums is not None:
    436     tape.trainable_params = argnums
--> 437 tapes, processing_fn = self.transform_fn(tape, *targs, **tkwargs)
    439 if processing_fn is None:
    441     def processing_fn(x):

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/pulse_gradient.py:539, in _stoch_pulse_grad(tape, argnum, num_split_times, sampler_seed, shots, use_broadcasting)
    536 sampler_seed = sampler_seed or np.random.randint(18421)
    537 key = jax.random.PRNGKey(sampler_seed)
--> 539 return _expval_stoch_pulse_grad(tape, argnum, num_split_times, key, shots, use_broadcasting)

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/pulse_gradient.py:592, in _expval_stoch_pulse_grad(tape, argnum, num_split_times, key, shots, use_broadcasting)
    589     continue
    591 key, _key = jax.random.split(key)
--> 592 cjacs, _tapes, avg_prefactor = _generate_tapes_and_cjacs(
    593     tape, idx, _key, num_split_times, use_broadcasting
    594 )
    596 gradient_data.append((len(_tapes), qml.math.stack(cjacs), avg_prefactor))
    597 tapes.extend(_tapes)

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/pulse_gradient.py:564, in _generate_tapes_and_cjacs(tape, idx, key, num_split_times, use_broadcasting)
    562 t0, *_, t1 = op.t
    563 taus = jnp.sort(jax.random.uniform(key, shape=(num_split_times,)) * (t1 - t0) + t0)
--> 564 cjacs = [cjac_fn(op.data[term_idx], tau) for tau in taus]
    565 if use_broadcasting:
    566     split_evolve_ops = _split_evol_ops(op, word, ob.wires, taus)

File ~/Qottmann/Xanadu/pennylane/pennylane/gradients/pulse_gradient.py:564, in (.0)
    562 t0, *_, t1 = op.t
    563 taus = jnp.sort(jax.random.uniform(key, shape=(num_split_times,)) * (t1 - t0) + t0)
--> 564 cjacs = [cjac_fn(op.data[term_idx], tau) for tau in taus]
    565 if use_broadcasting:
    566     split_evolve_ops = _split_evol_ops(op, word, ob.wires, taus)

    [... skipping hidden 7 frame]

File ~/Qottmann/Xanadu/pennylane/pennylane/pulse/transmon.py:499, in AmplitudeAndPhaseAndFreq.__call__(self, params, t)
    498 def __call__(self, params, t):
--> 499     return self.func(params, t)

File ~/Qottmann/Xanadu/pennylane/pennylane/pulse/transmon.py:469, in AmplitudeAndPhaseAndFreq.__init__..callable_amp(params, t)
    468 def callable_amp(params, t):
--> 469     return hz_to_rads * amp(params[0], t) * trig_fn(phase + hz_to_rads * freq * t)

Cell In[20], line 17, in amp(p, t)
     16 def amp(p, t):
---> 17     return jnp.polyval(p, t)

    [... skipping hidden 12 frame]

File ~/anaconda3/envs/pennylane311/lib/python3.11/site-packages/jax/_src/numpy/polynomial.py:231, in polyval(p, x, unroll)
    229 shape = lax.broadcast_shapes(p.shape[1:], x.shape)
    230 y = lax.full_like(x, 0, shape=shape, dtype=x.dtype)
--> 231 y, _ = lax.scan(lambda y, p: (y * x + p, None), y, p, unroll=unroll)
    232 return y

    [... skipping hidden 2 frame]

File ~/anaconda3/envs/pennylane311/lib/python3.11/site-packages/jax/_src/lax/control_flow/loops.py:189, in (.0)
    186 xs_flat, xs_tree = tree_flatten(xs)
    188 try:
--> 189   lengths = [x.shape[0] for x in xs_flat]
    190 except AttributeError as err:
    191   msg = "scan got value with no leading axis to scan over: {}."

IndexError: tuple index out of range
```

### System information

```shell
latest master, python 3.11
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by dwierichs** [comment]
This is a tricky problem to solve: 
1. The reordering in `HardwareHamiltonian.reorder_fn` only happens when calling the Hamiltonian, not at instantiation.
2. It is not possible to reorder the parameters at instantiation of the pulse (e.g. via `qml.evolve`) because the `Operation` class, to which `ParametrizedEvolution` has to adhere, requires `data` to be a list of tensors, not a nested list/tuple of tensors. At the same time, the `ParametrizedHamiltonian` class requires pulse envelopes to only take one parameter, so that the reordered parameters returned by `reorder_fn` need to be stacked into a list, rather than just copying the entries of the `data` list.
3. Because of all of the above, there was a preliminary decision to allow `HardwareHamiltonian` to leak into `stoch_pulse_grad`. Then we may compute the Jacobian of the `reorder_fn` and take the function and its Jacobian into account in the differentiation. However, it turns out that this is _very_ messy to implement because of the flexible parameter signatures allowed for pulses and the mixing of reordered and non-reordered parameters, all of which needs to be extracted from the behaviour of `reorder_fn`.

**Comment by dwierichs** [comment]
Resolved by #4215
